feat(tactic/lint): Three new linters, update illegal_constants

fpvandoorn opened the pull request:
* Implements two of the linters from #1924.
* Updates `illegal_constants` to only disallow `ge`/`gt` in the conclusion of declarations.
* Tries to fix newly introduced failures of the linter (likely I introduced some errors).

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [x] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [x] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

fpvandoorn committed:
add three new linters

fpvandoorn committed:
fix failing declarations

fpvandoorn committed:
restrict and rename illegal_constants linter

fpvandoorn committed:
update doc

fpvandoorn committed:
update ge_or_gt test

fpvandoorn committed:
update mk_nolint

robertylewis commented on scripts/mk_nolint.lean:
```diff
@@ -25,7 +25,8 @@ open io io.fs
 meta def active_linters :=
 [`linter.unused_arguments, `linter.dup_namespace, `linter.doc_blame,
- `linter.illegal_constants, `linter.def_lemma, `linter.instance_priority]
+ `linter.ge_or_gt, `linter.def_lemma, `linter.instance_priority, `linter.has_inhabited_instance
```
```suggestion
 `linter.ge_or_gt, `linter.def_lemma, `linter.instance_priority, `linter.has_inhabited_instance,
```

fpvandoorn committed:
fix error

robertylewis commented on src/meta/expr.lean:
```diff
@@ -439,6 +445,15 @@ meta def is_default_local : expr → bool
 | (expr.local_const _ _ binder_info.default _) := tt
 | _ := ff
 
+/-- Checks whether local constant `l` occurs in expression `e` -/
```
```suggestion
/-- `has_local_constant e l` checks whether local constant `l` occurs in expression `e` -/
```

robertylewis commented on src/tactic/lint.lean:
```diff
@@ -361,6 +377,70 @@ meta def linter.has_inhabited_instance : linter :=
+succeeds $ has_attribute `class nm <|> do
+  d ← get_decl nm,
+  tt ← unfolds_to_class d.value.lambda_body.pi_codomain.get_app_fn.const_name,
+  return 0
```
It took me a minute to parse what's going on here -- `tactic bool` ending with `return 0` is confusing. Maybe using parens instead of `$` would make this easier to read?

fpvandoorn replied:
added a comment

robertylewis reviewed (commented):
Nice!

fpvandoorn committed:
Update scripts/mk_nolint.lean

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

fpvandoorn committed:
Update src/meta/expr.lean

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

fpvandoorn committed:
clarify unfolds_to_class

fpvandoorn committed:
fix names since name is no longer protected

also change one declaration back to instance, since it did not cause a linter failure

robertylewis commented on src/category/monad/writer.lean:
```diff
@@ -148,7 +148,8 @@ export monad_writer_adapter (adapt_writer)
 variables {ω ω' : Type u} {m m' : Type u → Type v}
 
-@[priority 100] -- see Note [lower instance priority]
+/- marked as nolint, because it generates metavariables during type class resolution -/
```
Is this a problem? I suppose there aren't many relevant instances here to create bad searches.

fpvandoorn replied:
This instance is currently not dangerous. I looked at all the instances for these classes, and the search is finite, despite the metavariables. It might become problematic once we add recursive instances (see also my next comment).

robertylewis commented on src/category_theory/concrete_category/bundled_hom.lean:
```diff
@@ -57,7 +58,8 @@ intros; apply 𝒞.hom_ext;
 
 /-- A category given by `bundled_hom` is a concrete category. -/
-instance concrete_category : concrete_category (bundled c) :=
+-- The linter `dangerous_instance` fails, because it generates the type-class problem bundled_hom ?m
```
Again, I can't tell if this is truly dangerous or not without knowing the library better. My gut feeling is that if we sometimes have to ignore the linter "because here it's okay," the linter test probably isn't right. But if here it's okay just by circumstance, maybe we should trust the linter and this shouldn't be an instance?

fpvandoorn replied:
In this case it is okay because there are 0 instances for the `bundled_hom` class, so it will only check local constants. 

It will be problematic if there are any `bundled_hom`-instances that are recursive, i.e. that have type-class arguments themselves (in a way that can result in an infinite loop).

I guess I could test for this, but it sounds pretty tricky.

robertylewis replied:
Hmm. There are a few instances of `bundled_hom` (https://leanprover-community.github.io/mathlib_docs/category_theory/concrete_category/bundled_hom.html#category_theory.bundled_hom), but nothing that will make it loop.

Maybe we should move the warning into the docstring to make it more visible (here and above)? These are potentially bad instances. If the "wrong" good instances get added to the library, these could cause issues. It seems worth it to flag that as clearly as possible.

urkud replied:
A possible solution is to include `hom` type in the `bundled` signature.

fpvandoorn replied:
> There are a few instances of `bundled_hom`

I might have accidentally checked in the file `bundled_hom` instead of a file that imports all of mathlib, but I don't see any instances of the class `bundled_hom` in the file you linked.

I'm happy to add the warning to the docstring. These instances could indeed become dangerous if we do seemingly-innocuous things.

fpvandoorn replied:
Oh, in the file `unbundled_hom` there is an instance for `bundled_hom`.

fpvandoorn replied:
Oh! It lists all instances just below the class. Nice!

fpvandoorn replied:
@urkud: adding `hom` as a field instead of an argument would definitely solve these false positives, and that seems like a good thing to do. I don't know whether it is a problem that `bundled_hom.hom` is not syntactically equal to whatever it is defined as.

urkud replied:
@fpvandoorn I meant adding `hom` as an unused argument to the `bundled` function.

fpvandoorn replied:
Oh, sorry, you said adding it as an argument to `bundled` (I misread it as `bundled_hom`). Doesn't that run into issues if you need to synthesize `category (bundled c ?m)`? I don't know, maybe that works.

robertylewis commented on src/meta/expr.lean:
```diff
@@ -382,6 +382,12 @@ meta def pi_codomain : expr → expr -- see note [open expressions]
 
+/-- Get the body/value of a lambda-expression.
+  This definition doesn't instantiate bound variables, and therefore produces a term that is open. -/
+meta def lambda_body : expr → expr -- see note [open expressions]
```
If you move `Note [open expressions]` to the doc string it will get linkified in the html docs.

fpvandoorn committed:
fix errors, move notes to docstrings

fpvandoorn committed:
add comments to docstring

fpvandoorn committed:
update mk_all.sh

fpvandoorn committed:
fix linter errors

mergify[bot] committed:
Merge branch 'master' into new_linters
